// si-tani-ntt-frontend/src/components/CommodityChoroplethMap.jsx
import { useEffect, useMemo } from "react";
import { CircleMarker, LayersControl, MapContainer, Popup, TileLayer, useMap } from "react-leaflet";
import L from "leaflet";
import { AlertCircle, MapPin } from "lucide-react";

function getColor(value, max) {
  if (!value || !max) return "#94a3b8";
  const ratio = value / max;
  if (ratio > 0.85) return "#14532d";
  if (ratio > 0.65) return "#166534";
  if (ratio > 0.45) return "#15803d";
  if (ratio > 0.25) return "#16a34a";
  if (ratio > 0.1) return "#4ade80";
  return "#bbf7d0";
}

function bubbleRadius(value, max, selected = false) {
  if (!value || !max) return selected ? 10 : 8;
  const ratio = value / max;
  const base = 8 + Math.round(ratio * 18);
  return selected ? base + 4 : base;
}

function formatNumber(value, digits = 2) {
  return new Intl.NumberFormat("id-ID", { maximumFractionDigits: digits }).format(Number(value || 0));
}

function fitBoundsFromPoints(points = []) {
  const valid = points.filter((item) => Number.isFinite(Number(item.latitude)) && Number.isFinite(Number(item.longitude)));
  if (!valid.length) return null;
  return L.latLngBounds(valid.map((item) => [Number(item.latitude), Number(item.longitude)]));
}

function FitToMarkers({ points = [], selectedRegencyCode }) {
  const map = useMap();

  useEffect(() => {
    if (!map || !points.length) return;
    const selected = points.find((item) => item.regencyCode === selectedRegencyCode);
    if (selected && Number.isFinite(Number(selected.latitude)) && Number.isFinite(Number(selected.longitude))) {
      map.flyTo([Number(selected.latitude), Number(selected.longitude)], Math.max(map.getZoom(), 9), { duration: 0.8 });
      return;
    }
    const bounds = fitBoundsFromPoints(points);
    if (bounds?.isValid()) map.fitBounds(bounds.pad(0.14));
  }, [map, points, selectedRegencyCode]);

  return null;
}

function Legend({ maxValue, unitLabel }) {
  const steps = [0.1, 0.25, 0.45, 0.65, 0.85].map((ratio) => ({
    ratio,
    value: maxValue * ratio,
    color: getColor(maxValue * ratio, maxValue)
  }));

  return (
    <div className="absolute bottom-4 left-4 z-[500] w-56 rounded-2xl border border-slate-200 bg-white/95 p-3 shadow-lg backdrop-blur">
      <div className="text-xs font-semibold uppercase tracking-[0.14em] text-slate-400">Legenda</div>
      <div className="mt-2 text-sm font-semibold text-slate-900">Intensitas indikator per wilayah</div>
      <div className="mt-1 text-[11px] leading-5 text-slate-500">Mode aman memakai titik centroid kabupaten/kota agar tidak tertukar oleh boundary offline lama.</div>
      <div className="mt-3 space-y-2 text-xs text-slate-600">
        {steps.reverse().map((step) => (
          <div key={step.ratio} className="flex items-center justify-between gap-3">
            <div className="flex items-center gap-2">
              <span className="h-3 w-3 rounded-full" style={{ backgroundColor: step.color }} />
              <span>{step.ratio === 0.85 ? "Sangat tinggi" : step.ratio === 0.65 ? "Tinggi" : step.ratio === 0.45 ? "Menengah" : step.ratio === 0.25 ? "Rendah" : "Sangat rendah"}</span>
            </div>
            <span>{formatNumber(step.value, 1)} {unitLabel}</span>
          </div>
        ))}
      </div>
    </div>
  );
}

export default function CommodityChoroplethMap({ regencies = [], metricLabel = "Produksi", unitLabel = "ton", selectedRegencyCode = null, onSelectRegency, height = "560px", note = "" }) {
  const maxValue = useMemo(() => Math.max(...regencies.map((item) => Number(item.value || 0)), 0), [regencies]);
  const points = useMemo(
    () => regencies.filter((item) => Number.isFinite(Number(item.latitude)) && Number.isFinite(Number(item.longitude))),
    [regencies]
  );

  if (!points.length) {
    return <div className="flex h-full min-h-[420px] items-center justify-center rounded-3xl border border-dashed border-slate-200 bg-slate-50 text-sm text-slate-500">Koordinat wilayah belum tersedia untuk peta komoditas.</div>;
  }

  return (
    <div className="relative overflow-hidden rounded-[28px] border border-slate-200" style={{ height }}>
      <MapContainer center={[-8.7, 121.1]} zoom={8} className="h-full w-full" preferCanvas scrollWheelZoom>
        <LayersControl position="topright">
          <LayersControl.BaseLayer checked name="Satelit">
            <TileLayer attribution='Tiles &copy; Esri' url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}" />
          </LayersControl.BaseLayer>
          <LayersControl.BaseLayer name="Peta standar">
            <TileLayer attribution='&copy; OpenStreetMap contributors' url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png" />
          </LayersControl.BaseLayer>
          <LayersControl.Overlay checked name="Label wilayah">
            <TileLayer attribution='Tiles &copy; Esri' url="https://services.arcgisonline.com/ArcGIS/rest/services/Reference/World_Boundaries_and_Places/MapServer/tile/{z}/{y}/{x}" />
          </LayersControl.Overlay>
        </LayersControl>
        <FitToMarkers points={points} selectedRegencyCode={selectedRegencyCode} />
        {points.map((item) => {
          const selected = item.regencyCode === selectedRegencyCode;
          const radius = bubbleRadius(item.value, maxValue, selected);
          const color = getColor(item.value, maxValue);
          return (
            <CircleMarker
              key={item.regencyCode}
              center={[Number(item.latitude), Number(item.longitude)]}
              radius={radius}
              pathOptions={{
                color: selected ? "#0f172a" : color,
                weight: selected ? 2.8 : 1.8,
                fillColor: color,
                fillOpacity: item.hasData ? 0.88 : 0.45
              }}
              eventHandlers={{ click: () => onSelectRegency?.(item.regencyCode) }}
            >
              <Popup>
                <div className="min-w-[230px] space-y-1">
                  <div className="font-semibold text-slate-900">{item.regencyFullName || item.regencyName}</div>
                  <div className="text-xs text-slate-500">Titik wilayah kabupaten/kota untuk analitik komoditas.</div>
                  <div className="pt-2 text-xs text-slate-600">Nilai indikator: <strong>{formatNumber(item.value, 2)} {item.unit || unitLabel}</strong></div>
                  <div className="text-xs text-slate-600">Komoditas utama: <strong>{item.topCommodity || "-"}</strong></div>
                  <div className="text-xs text-slate-600">Dataset: <strong>{(item.datasets || []).slice(0, 3).join(", ") || "Belum ada data"}</strong></div>
                  <div className="text-xs text-slate-600">Cakupan record: <strong>{item.coverage || 0}</strong></div>
                </div>
              </Popup>
            </CircleMarker>
          );
        })}
        <Legend maxValue={maxValue} unitLabel={unitLabel} />
      </MapContainer>
      <div className="absolute right-4 top-4 z-[500] max-w-sm rounded-2xl border border-slate-200 bg-white/95 px-4 py-3 shadow-lg backdrop-blur">
        <div className="text-xs font-semibold uppercase tracking-[0.14em] text-slate-400">Peta analitik komoditas</div>
        <div className="mt-1 text-sm font-semibold text-slate-900">Klik titik wilayah kabupaten/kota untuk drilldown {metricLabel.toLowerCase()}</div>
        <div className="mt-2 flex items-start gap-2 text-xs leading-5 text-slate-500"><MapPin size={14} className="mt-0.5 shrink-0 text-emerald-600" />Mode peta saat ini memakai centroid wilayah agar analisis tetap akurat selama boundary offline lama belum diganti dengan boundary resmi terbaru.</div>
        {note ? <div className="mt-2 flex items-start gap-2 text-xs leading-5 text-slate-500"><AlertCircle size={14} className="mt-0.5 shrink-0 text-amber-500" />{note}</div> : null}
      </div>
    </div>
  );
}
